wave_kinematics/src/quadrotor.cpp

```diff
@@ -0,0 +1,435 @@
+                                   Vec4 actual,
+                                   double dt) {
+    Vec4 setpoints;
+    setpoints = psetpoints + vsetpoints;
+    return this->update(setpoints, actual, dt);
+}
+
+
+// POSITION CONTROLLER
+PositionController::PositionController(void) {
+    this->dt = 0.0;
+    this->outputs << 0.0, 0.0, 0.0, 0.0;
+
+    this->x_controller = PID(0.5, 0.0, 0.035);
+    this->y_controller = PID(0.5, 0.0, 0.035);
+    this->z_controller = PID(0.3, 0.0, 0.018);
+}
+
+Vec4 PositionController::update(Vec3 setpoints,
+                                   Vec4 actual,
+                                   double yaw,
+                                   double dt) {
+    double r, p, y, t;
+    Vec3 errors, euler;
+    Vec4 outputs;
+    Mat3 R;
+
+    // check rate
+    this->dt += dt;
+    if (this->dt < 0.01) {
+        return this->outputs;
+    }
+
+    // calculate RPY errors relative to quadrotor by incorporating yaw
+    errors(0) = setpoints(0) - actual(0);
+    errors(1) = setpoints(1) - actual(1);
+    errors(2) = setpoints(2) - actual(2);
+    euler << 0.0, 0.0, actual(3);
+    euler2rot(euler, 123, R);
+    errors = R * errors;
+
+    // roll, pitch, yaw and thrust
+    r = -this->y_controller.update(errors(1), 0.0, dt);
+    p = this->x_controller.update(errors(0), 0.0, dt);
+    y = yaw;
+    t = 0.5 + this->z_controller.update(errors(2), 0.0, dt);
+    outputs << r, p, y, t;
+
+    // limit roll, pitch
+    for (int i = 0; i < 2; i++) {
+        if (outputs(i) > deg2rad(30.0)) {
+            outputs(i) = deg2rad(30.0);
+        } else if (outputs(i) < deg2rad(-30.0)) {
+            outputs(i) = deg2rad(-30.0);
+        }
+    }
+
+    // limit yaw
+    while (outputs(2) > deg2rad(360.0)) {
+        outputs(2) -= deg2rad(360.0);
+    }
+    while (outputs(2) < deg2rad(0.0)) {
+        outputs(2) += deg2rad(360.0);
+    }
+
+    // limit thrust
+    if (outputs(3) > 1.0) {
+        outputs(3) = 1.0;
+    } else if (outputs(3) < 0.0) {
+        outputs(3) = 0.0;
+    }
+
+    // yaw first if threshold reached
+    if (fabs(yaw - actual(3)) > deg2rad(2)) {
+        outputs(0) = 0.0;
+        outputs(1) = 0.0;
+    }
+
+    // keep track of outputs
+    this->outputs = outputs;
+    this->dt = 0.0;
+
+    return outputs;
+}
+
+
+// QUADROTOR MODEL
+QuadrotorModel::QuadrotorModel(void) {
+    this->states = VecX(12);
+    this->states(0) = 0.0;   // roll
+    this->states(1) = 0.0;   // pitch
+    this->states(2) = 0.0;   // yaw
+    this->states(3) = 0.0;   // angular velocity x
+    this->states(4) = 0.0;   // angular velocity y
+    this->states(5) = 0.0;   // angular velocity z
+    this->states(6) = 0.0;   // x
+    this->states(7) = 0.0;   // y
+    this->states(8) = 0.0;   // z
+    this->states(9) = 0.0;   // linear velocity x
+    this->states(10) = 0.0;  // linear velocity y
+    this->states(11) = 0.0;  // linear velocity z
+
+    this->Ix = 0.0963;  // inertial x
```

Comment:
Put these comments into the header (or, use non-abbreviated variable names).
Put the initialization into the header as well.
Some of these should be `const`.